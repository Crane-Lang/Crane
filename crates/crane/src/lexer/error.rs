use serde::{Serialize, Deserialize};
use thiserror::Error;

use crate::ast::Span;

#[derive(Error, Debug, PartialEq, Clone, Serialize, Deserialize)]
pub struct LexError {
    pub kind: LexErrorKind,
    pub span: Span,
}

impl Default for LexError {
    fn default() -> Self {
        Self {
            kind: Default::default(),
            span: Span::new(0, 0),
        }
    }
}

impl std::fmt::Display for LexError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.kind)
    }
}

#[derive(Debug, Error, PartialEq, Clone, Default, Serialize, Deserialize)]
pub enum LexErrorKind {
    #[default]
    #[error("unknown")]
    Unknown,
}
