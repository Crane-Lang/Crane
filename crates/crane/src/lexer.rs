mod error;

pub mod token;

pub use error::*;

use logos::Logos;

use crate::lexer::token::{Token, TokenKind};

pub struct Lexer<'src> {
    lexer: logos::Lexer<'src, TokenKind>,
}

impl<'src> Lexer<'src> {
    pub fn new(input: &'src str) -> Self {
        Self {
            lexer: TokenKind::lexer(input),
        }
    }
}

impl<'src> Iterator for Lexer<'src> {
    type Item = Result<Token, LexError>;

    fn next(&mut self) -> Option<Self::Item> {
        let kind = self.lexer.next()?;
        let lexeme = self.lexer.slice();
        let span = self.lexer.span().into();

        Some(
            kind.map(|kind| Token {
                kind,
                lexeme: lexeme.into(),
                span,
            })
            .map_err(|kind| LexError { kind, span }),
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn lex(input: &str) -> logos::Lexer<TokenKind> {
        TokenKind::lexer(input)
    }

    fn check(input: &str, token: TokenKind) {
        let mut lexer = TokenKind::lexer(input);

        assert_eq!(lexer.next(), Some(Ok(token)));
        assert_eq!(lexer.slice(), input);
    }

    #[test]
    fn lex_open_paren() {
        check("(", TokenKind::OpenParen)
    }

    #[test]
    fn lex_close_paren() {
        check(")", TokenKind::CloseParen)
    }

    #[test]
    fn lex_open_brace() {
        check("{", TokenKind::OpenBrace)
    }

    #[test]
    fn lex_close_brace() {
        check("}", TokenKind::CloseBrace)
    }

    #[test]
    fn lex_comment() {
        check("// This is a comment.", TokenKind::Comment)
    }

    #[test]
    fn lex_fn_declaration() {
        let mut lex = lex("fn foo() {}");

        assert_eq!(lex.next(), Some(Ok(TokenKind::Ident)));
        assert_eq!(lex.span(), 0..2);
        assert_eq!(lex.slice(), "fn");

        assert_eq!(lex.next(), Some(Ok(TokenKind::Ident)));
        assert_eq!(lex.span(), 3..6);
        assert_eq!(lex.slice(), "foo");

        assert_eq!(lex.next(), Some(Ok(TokenKind::OpenParen)));
        assert_eq!(lex.span(), 6..7);
        assert_eq!(lex.slice(), "(");

        assert_eq!(lex.next(), Some(Ok(TokenKind::CloseParen)));
        assert_eq!(lex.span(), 7..8);
        assert_eq!(lex.slice(), ")");

        assert_eq!(lex.next(), Some(Ok(TokenKind::OpenBrace)));
        assert_eq!(lex.span(), 9..10);
        assert_eq!(lex.slice(), "{");

        assert_eq!(lex.next(), Some(Ok(TokenKind::CloseBrace)));
        assert_eq!(lex.span(), 10..11);
        assert_eq!(lex.slice(), "}");
    }

    #[test]
    fn test_lexer() {
        insta::glob!("snapshot_inputs/*.crane", |path| {
            let source = std::fs::read_to_string(path).unwrap();

            let lexer = Lexer::new(&source);

            insta::assert_yaml_snapshot!(lexer.into_iter().collect::<Vec<_>>());
        })
    }
}
