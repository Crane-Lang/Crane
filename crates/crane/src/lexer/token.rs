use logos::Logos;
use smol_str::SmolStr;

use crate::ast::Span;
use crate::lexer::LexErrorKind;

#[derive(Logos, Debug, PartialEq, Clone, Copy)]
#[logos(error = LexErrorKind)]
pub enum TokenKind {
    /// `(`
    #[token("(")]
    OpenParen,

    /// `)`
    #[token(")")]
    CloseParen,

    /// `{`
    #[token("{")]
    OpenBrace,

    /// `}`
    #[token("}")]
    CloseBrace,

    /// `,`
    #[token(",")]
    Comma,

    /// An identifier.
    #[regex(r"[A-Za-z_]+")]
    Ident,

    /// A string literal.
    #[regex(r#""[^"]*""#)]
    String,

    /// An integer literal.
    #[regex(r#"[\d]+"#)]
    Integer,

    /// A comment.
    #[regex(r"//.*")]
    Comment,

    /// Any sequence of whitespace characters.
    #[regex(r"[ \n]+", logos::skip)]
    Whitespace,
}

#[derive(Debug, Clone)]
pub struct Token {
    pub kind: TokenKind,
    pub lexeme: SmolStr,
    pub span: Span,
}
