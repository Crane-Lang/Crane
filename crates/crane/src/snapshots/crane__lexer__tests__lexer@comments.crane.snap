---
source: crates/crane/src/lexer.rs
expression: "lexer.into_iter().collect::<Vec<_>>()"
input_file: crates/crane/src/snapshot_inputs/comments.crane
---
- Ok:
    kind: Ident
    lexeme: use
    span:
      start: 0
      end: 3
- Ok:
    kind: Ident
    lexeme: std
    span:
      start: 4
      end: 7
- Ok:
    kind: ColonColon
    lexeme: "::"
    span:
      start: 7
      end: 9
- Ok:
    kind: Ident
    lexeme: io
    span:
      start: 9
      end: 11
- Ok:
    kind: ColonColon
    lexeme: "::"
    span:
      start: 11
      end: 13
- Ok:
    kind: Ident
    lexeme: println
    span:
      start: 13
      end: 20
- Ok:
    kind: Ident
    lexeme: fn
    span:
      start: 22
      end: 24
- Ok:
    kind: Ident
    lexeme: main
    span:
      start: 25
      end: 29
- Ok:
    kind: OpenParen
    lexeme: (
    span:
      start: 29
      end: 30
- Ok:
    kind: CloseParen
    lexeme: )
    span:
      start: 30
      end: 31
- Ok:
    kind: OpenBrace
    lexeme: "{"
    span:
      start: 32
      end: 33
- Ok:
    kind: Ident
    lexeme: println
    span:
      start: 38
      end: 45
- Ok:
    kind: OpenParen
    lexeme: (
    span:
      start: 45
      end: 46
- Ok:
    kind: Ident
    lexeme: always_blue
    span:
      start: 46
      end: 57
- Ok:
    kind: OpenParen
    lexeme: (
    span:
      start: 57
      end: 58
- Ok:
    kind: CloseParen
    lexeme: )
    span:
      start: 58
      end: 59
- Ok:
    kind: CloseParen
    lexeme: )
    span:
      start: 59
      end: 60
- Ok:
    kind: CloseBrace
    lexeme: "}"
    span:
      start: 61
      end: 62
- Ok:
    kind: DocComment
    lexeme: "/// Always returns the string \"blue\"."
    span:
      start: 64
      end: 101
- Ok:
    kind: Ident
    lexeme: fn
    span:
      start: 102
      end: 104
- Ok:
    kind: Ident
    lexeme: always_blue
    span:
      start: 105
      end: 116
- Ok:
    kind: OpenParen
    lexeme: (
    span:
      start: 116
      end: 117
- Ok:
    kind: CloseParen
    lexeme: )
    span:
      start: 117
      end: 118
- Ok:
    kind: RightArrow
    lexeme: "->"
    span:
      start: 119
      end: 121
- Ok:
    kind: Ident
    lexeme: String
    span:
      start: 122
      end: 128
- Ok:
    kind: OpenBrace
    lexeme: "{"
    span:
      start: 129
      end: 130
- Ok:
    kind: Comment
    lexeme: // Always blue! Always blue!
    span:
      start: 135
      end: 163
- Ok:
    kind: String
    lexeme: "\"blue\""
    span:
      start: 168
      end: 174
- Ok:
    kind: CloseBrace
    lexeme: "}"
    span:
      start: 175
      end: 176

